src/bootloader/bootloader.cpp

```diff
@@ -0,0 +1,286 @@
+///
+template <typename ProgressCallback>
+static void copy_bootloader_to_flash(FILE *bootloader_bin, ProgressCallback progress) {
+    const size_t total_bytes = 131072;
+    size_t bytes_in_preceding_sectors = 0;
+
+    for (unsigned sector = 0; sector < bootloader_sector_count; sector++) {
+
+        // do not reflash preboot if not necessary
+        if (sector == 0) {
+            uint32_t expected_preboot_crc = 0;
+            if (!calculate_file_crc(bootloader_bin, bootloader_sector_get_size(0), expected_preboot_crc)) {
+                fatal_error("expected preboot crc calculation failed");
+            }
+
+            uint32_t current_preboot_crc = crc32_calc(bootloader_sector_get_address(0), bootloader_sector_get_size(0));
+            if (current_preboot_crc == expected_preboot_crc && false) {
+                log(LOG_SEVERITY_INFO, "No need to update preboot. Skipping sector 0.");
+                continue;
+            } else {
+                log(LOG_SEVERITY_INFO, "Going to update preboot now.");
+            }
+        }
+
+        log(LOG_SEVERITY_INFO, "Flashing sector %i", sector);
+
+        // add random delay to make preboot flashing less predictable
+        if (sector == 0) {
+            uint32_t delay_ms = 100 + (random_number() % 7000);
+            osDelay(delay_ms);
+        }
+
+        // seek at the sector in the file
+        if (fseek(bootloader_bin, bootloader_sector_get_address(sector) - bootloader_sector_get_address(0), SEEK_SET) != 0) {
+            fatal_error("bootloader update: failed to seek sector");
+        }
+
+        // erase the sector
+        HAL_FLASH_Unlock();
+        if (!flash_erase_sector(sector)) {
+            fatal_error("bootloader update: failed to erase sector");
+        }
+
+        // program the sector
+        HAL_FLASH_Unlock();
+        bool flash_successful = flash_program_sector(sector, bootloader_bin, [&](size_t bytes_written) {
+            if (sector != 0) {
+                // do not report progress for sector 0, as updating preboot is potentially dangerous
+                // and we want to be as quick as possible and minimize the code running in-between
+                size_t total_written = bytes_in_preceding_sectors + bytes_written;
+                int percent_done = 100 * total_written / total_bytes;
+                progress(percent_done);
+            }
+        });
+        HAL_FLASH_Lock();
+
+        if (!flash_successful) {
+            fatal_error("bootloader update: failed to flash sector");
+        }
+
+        log(LOG_SEVERITY_INFO, "Sector %i flashed successfully", sector);
+        bytes_in_preceding_sectors += bootloader_sector_get_size(sector);
+    }
+}
+
+Version buddy::bootloader::get_version() {
+    return *bootloader_version;
+}
+
+bool buddy::bootloader::needs_update() {
+    auto current = get_version();
+    auto required = buddy::bootloader::required_version;
+
+    return current.major < required.major
```

Comment:
ideally rewrite this using `std::tie` + `operator<` thereof ... I always make errors in these comparisons of sequences

Reply:
Great idea 🙏